### PR TITLE
Manually filter for nearby overlapping structures for Mex Cap

### PR DESCRIPTION
**Pull request by aeoncleanse**
This _should_ fix the issue we've been having with the engine for some reason letting the mex cap storages overlap other structures. Needs THOROUGH testing!!!!!

### CODE REVIEW

**Comment by PhilipJFryFAF** [comment]
tested and doesn't fix the issue

[replay-id](replay.faforever.com/6031267)

**Comment by TheKeyBlue** [comment]
I checked the code and that looks good to me.
I also did some ingame tests and was unable to reproduce the bug using the same methods I can currently use in the develop version.

File: lua/utilities.lua [hunk 1]
```diff
@@ -9,6 +9,56 @@ function GetDistanceBetweenTwoEntities(entity1, entity2)
...
+            local zDist = math.abs(pos.z - structPos.z)
+
+            local skirtDiffx = mySkirtX + (structPhysics.SkirtSizeX / 2)
+            local skirtDiffz = mySkirtZ + (structPhysics.SkirtSizeZ / 2)
+
+            -- Check if the axis difference is smaller than the combined skirt distance
+            -- If it is, we overlap, and can't build here
+            if xDist < skirtDiffx and zDist < skirtDiffz then
+                allowed = false
+                break
```
Comment (HUSSAR-mtrela) [inline lua/utilities.lua]: you could just `return false` here

File: lua/utilities.lua [hunk 2]
```diff
@@ -9,6 +9,56 @@ function GetDistanceBetweenTwoEntities(entity1, entity2)
...
+    local x1 = pos.x - xDiff
+    local z1 = pos.z - zDiff
+    local x2 = pos.x + xDiff
+    local z2 = pos.z + zDiff
+
+    -- Find all the units in that rectangle
+    local units = GetUnitsInRect(Rect(x1, z1, x2, z2))
+
+    -- Filter it down to structures only
+    units = EntityCategoryFilterDown(categories.STRUCTURE, units)
```
Comment (HUSSAR-mtrela) [inline lua/utilities.lua]: IMO, you should also check for units that can be build outside of factories. For example. building Megalith next to mex should prevent building of a mass storage. In other words filtered `units `table should contain `STRUCTURES` or `EXPERIMENTAL` units

Comment (aeoncleanse) [inline lua/utilities.lua]: That would be categories.STRUCTURE + categories.EXPERIMENTAL right ?

Comment (speed2CZ) [inline lua/utilities.lua]: yes

Comment (HUSSAR-mtrela) [inline lua/utilities.lua]: yes, that's correct.

**Pull request merged**